graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -156,18 +160,47 @@ torch::Tensor UVAIndexSelectImpl_(torch::Tensor input, torch::Tensor index) {
 /**
  * @brief UVA index select operator implementation on CUDA.
  *
- * The supporting input types are: float, double, int, int64_t.
+ * All basic torch types are supported for input.
  * The supporting index types are: int, int64_t.
  */
 torch::Tensor UVAIndexSelectImpl(torch::Tensor input, torch::Tensor index) {
-  return AT_DISPATCH_FLOATING_TYPES_AND2(
-      at::ScalarType::Int, at::ScalarType::Long, input.scalar_type(),
-      "UVAIndexSelectImpl", [&] {
-        return AT_DISPATCH_INDEX_TYPES(
-            index.scalar_type(), "UVAIndexSelectImpl", [&] {
-              return UVAIndexSelectImpl_<scalar_t, index_t>(input, index);
-            });
-      });
+  return AT_DISPATCH_INDEX_TYPES(
+      index.scalar_type(), "UVAIndexSelectImpl", ([&] {
+        const auto ptr = (size_t)input.data_ptr();
+        const int64_t feature_size = std::accumulate(
+            input.sizes().begin() + 1, input.sizes().end(), 1ll,
+            std::multiplies<>());
+        // We perform the copy with datatype of size powers of 2, and the
+        // maximum data type we use has 16 bytes. We check the alignment of the
+        // pointer and the feature dimensionality to determine the largest
+        // type to use for the copy to minimize the number of CUDA threads used.
+        // Alignment denotes the maximum suitable alignment and datatype size
+        // for the copies.
+        const int alignment =
+            std::gcd(16, std::gcd(ptr, input.element_size() * feature_size));
+        const auto new_feature_size =
+            input.element_size() * feature_size / alignment;
+        switch (alignment) {
```

Comment:
I am not sure whether we need to support all these data types. If so, we need to add unit tests for them. @frozenbugs, what do you think?

Reply:
Yes, it is better to start from what we really need, if we foresee those are important for our future work, we can add it and unit test is required in this case.

Reply:
Let me add the unit tests for all different types of sizes 1 2 4 8 and 16 from torch. I have the code already so easy to extend the unit test with more datatypes.

Reply:
I added the unit tests for all possible types of different sizes.